services/motion/explore/explore.go

```diff
@@ -0,0 +1,620 @@
+	if err != nil {
+		return false, err
+	}
+
+	// Create kinematic base
+	kb, err := ms.createKinematicBase(ctx, componentName, extra)
+	if err != nil {
+		return false, err
+	}
+
+	// Create motionplan plan
+	planInputs, err := ms.createMotionPlan(ctx, kb, destination.Pose(), worldState, true, extra)
+	if err != nil {
+		return false, err
+	}
+	var plan motionplan.Plan
+	for _, inputs := range planInputs {
+		input := make(map[string][]referenceframe.Input)
+		input[kb.Name().Name] = inputs
+		plan = append(plan, input)
+	}
+
+	// Start background processes
+	cancelCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// Start polling for obstacles
+	ms.backgroundWorkers.Add(1)
+	goutils.ManagedGo(func() {
+		ms.checkForObstacles(cancelCtx, obstacleDetectors, kb, plan, extra)
+	}, ms.backgroundWorkers.Done)
+
+	// Start executing plan
+	ms.backgroundWorkers.Add(1)
+	goutils.ManagedGo(func() {
+		ms.executePlan(cancelCtx, kb, plan)
+	}, ms.backgroundWorkers.Done)
+
+	for {
+		// this ensures that if the context is cancelled we always return early at the top of the loop
+		if err := ctx.Err(); err != nil {
+			return false, err
+		}
+
+		select {
+		// if context was cancelled by the calling function, error out
+		case <-ctx.Done():
+			return false, ctx.Err()
+
+		// once execution responds: return the result to the caller
+		case resp := <-ms.executionResponseChan:
+			ms.logger.Debugf("execution completed: %s", resp)
+			return resp.success, resp.err
+
+		// if the checkPartialPlan process hit an error return it, otherwise exit
+		case resp := <-ms.obstacleResponseChan:
+			ms.logger.Debugf("obstacle response: %s", resp)
+			if resp.err != nil {
+				return resp.success, resp.err
+			}
+			if resp.success {
+				return resp.success, nil
+			}
+		}
+	}
+}
+
+type moveResponse struct {
+	err     error
+	success bool
+}
+
+// checkForObstacles will continuously monitor the generated transient worldState for obstacles in the given
+// motionplan plan. A response will be sent through the channel if an error occurs, the motionplan plan
+// completes or an obstacle is detected in the given range.
+func (ms *explore) checkForObstacles(
+	ctx context.Context,
+	obstacleDetectors []obstacleDetectorPair,
+	kb kinematicbase.KinematicBase,
+	plan motionplan.Plan,
+	extra map[string]interface{},
+) {
+	// Get obstacle polling frequency from extra
+	var obstaclePollingFrequencyHz float64
+	if obstaclePollingFrequencyHzInterface, ok := extra["obstacle_polling_frequency_hz"]; ok {
+		obstaclePollingFrequencyHz, ok = obstaclePollingFrequencyHzInterface.(float64)
+		if !ok {
+			return
+		}
+	} else {
+		obstaclePollingFrequencyHz = defaultObstaclePollingFrequencyHz
+	}
+
+	// Constantly check for obstacles in path at desired obstacle polling frequency
+	ticker := time.NewTicker(time.Duration(int(1000/obstaclePollingFrequencyHz)) * time.Millisecond)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
```

Comment:
[note] I am still unclear when we should be sending to the obstacle channel and returning outside of the success case (obstacle detected within range). I believe this will become obvious after additional hardware testing, but leave  a note if you have any opinions